File: backend/src/tathya_avalokan/models/instance.py
import uuid
from typing import TYPE_CHECKING

from sqlalchemy import Boolean, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from tathya_avalokan.models.base import Base, TimestampMixin

if TYPE_CHECKING:
    from tathya_avalokan.models.project import Project


class DatabaseInstance(Base, TimestampMixin):
    """Internal metadata entity representing a configured database connection."""

    __tablename__ = "database_instances"

    id: Mapped[str] = mapped_column(
        String(36),
        primary_key=True,
        default=lambda: str(uuid.uuid4()),
    )
    project_id: Mapped[str] = mapped_column(
        String(36),
        ForeignKey("projects.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    driver_type: Mapped[str] = mapped_column(
        String(20),
        nullable=False,
        default="postgresql",
    )
    host: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )
    port: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )
    database_name: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )
    username: Mapped[str | None] = mapped_column(
        String(100),
        nullable=True,
    )
    # Symmetrically encrypted JSON string via Fernet
    encrypted_credentials: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    ssl_mode: Mapped[str] = mapped_column(
        String(30),
        nullable=False,
        default="prefer",
    )
    is_read_only: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=False,
    )

    # Relationship back to parent project
    project: Mapped["Project"] = relationship(
        "Project",
        back_populates="instances",
    )

File: backend/internal/models/instance.go
package models

import (
	"time"
)

type DriverType string

const (
	DriverPostgres DriverType = "postgresql"
	DriverMySQL    DriverType = "mysql"
	DriverSQLite   DriverType = "sqlite"
)

type DatabaseInstance struct {
	ID                   string     `json:"id"`
	ProjectID            string     `json:"project_id"`
	Name                 string     `json:"name"`
	DriverType           DriverType `json:"driver_type"`
	Host                 *string    `json:"host"`
	Port                 *int       `json:"port"`
	DatabaseName         string     `json:"database_name"`
	Username             *string    `json:"username"`
	EncryptedCredentials string     `json:"-"`
	SSLMode              string     `json:"ssl_mode"`
	IsReadOnly           bool       `json:"is_read_only"`
	CreatedAt            time.Time  `json:"created_at"`
	UpdatedAt            time.Time  `json:"updated_at"`
}

type DatabaseInstanceCreate struct {
	Name          string     `json:"name"`
	DriverType    DriverType `json:"driver_type"`
	Host          *string    `json:"host"`
	Port          *int       `json:"port"`
	DatabaseName  string     `json:"database_name"`
	Username      *string    `json:"username"`
	Password      *string    `json:"password"`
	ConnectionURL *string    `json:"connection_url"`
	SSLMode       *string    `json:"ssl_mode"`
	IsReadOnly    *bool      `json:"is_read_only"`
}

type DatabaseInstanceUpdate struct {
	Name          *string     `json:"name"`
	DriverType    *DriverType `json:"driver_type"`
	Host          *string     `json:"host"`
	Port          *int        `json:"port"`
	DatabaseName  *string     `json:"database_name"`
	Username      *string     `json:"username"`
	Password      *string     `json:"password"`
	ConnectionURL *string     `json:"connection_url"`
	SSLMode       *string     `json:"ssl_mode"`
	IsReadOnly    *bool       `json:"is_read_only"`
}

type DatabaseInstanceResponse struct {
	ID            string     `json:"id"`
	ProjectID     string     `json:"project_id"`
	Name          string     `json:"name"`
	DriverType    DriverType `json:"driver_type"`
	Host          *string    `json:"host"`
	Port          *int       `json:"port"`
	DatabaseName  string     `json:"database_name"`
	Username      *string    `json:"username"`
	SSLMode       string     `json:"ssl_mode"`
	IsReadOnly    bool       `json:"is_read_only"`
	IsPasswordSet bool       `json:"is_password_set"`
	CreatedAt     time.Time  `json:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at"`
}

func (inst *DatabaseInstance) ToResponse(isPasswordSet bool) DatabaseInstanceResponse {
	return DatabaseInstanceResponse{
		ID:            inst.ID,
		ProjectID:     inst.ProjectID,
		Name:          inst.Name,
		DriverType:    inst.DriverType,
		Host:          inst.Host,
		Port:          inst.Port,
		DatabaseName:  inst.DatabaseName,
		Username:      inst.Username,
		SSLMode:       inst.SSLMode,
		IsReadOnly:    inst.IsReadOnly,
		IsPasswordSet: isPasswordSet,
		CreatedAt:     inst.CreatedAt,
		UpdatedAt:     inst.UpdatedAt,
	}
}
